File: plugins/plugin-code/typescript/actions/writeFile.ts
import type {
  Action,
  ActionResult,
  HandlerCallback,
  HandlerOptions,
  IAgentRuntime,
  Memory,
  State,
} from "@elizaos/core";
import { requireActionSpec } from "../generated/specs/spec-helpers";
import type { CoderService } from "../services/coderService";

function getInputs(
  options: HandlerOptions | undefined,
  message: Memory,
): { filepath: string; content: string } {
  const opt = options as { filepath?: string; content?: string } | undefined;
  const filepath = opt?.filepath?.trim() ?? "";
  const content = opt?.content ?? "";

  if (filepath) return { filepath, content };

  const text = message.content.text ?? "";
  const m = text.match(/["'`]([^"'`]+)["'`]/);
  const inferred = m?.[1]?.trim() ?? "";
  return { filepath: inferred, content };
}

const spec = requireActionSpec("WRITE_FILE");

export const writeFile: Action = {
  name: spec.name,
  similes: spec.similes ? [...spec.similes] : [],
  description: spec.description,
  validate: async (runtime: IAgentRuntime): Promise<boolean> =>
    runtime.getService<CoderService>("coder") !== null,
  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    _state: State | undefined,
    options: HandlerOptions | undefined,
    callback?: HandlerCallback,
  ): Promise<ActionResult> => {
    const svc = runtime.getService<CoderService>("coder");
    if (!svc) {
      const msg = "Coder service is not available.";
      if (callback) await callback({ content: { text: msg } });
      return { success: false, text: msg };
    }

    const { filepath, content } = getInputs(options, message);
    if (!filepath) {
      const msg = "Missing filepath.";
      if (callback) await callback({ content: { text: msg } });
      return { success: false, text: msg };
    }

    const conv = message.roomId ?? message.agentId ?? runtime.agentId;
    const result = await svc.writeFile(conv, filepath, content);
    if (!result.ok) {
      if (callback) await callback({ content: { text: result.error } });
      return { success: false, text: result.error };
    }

    const msg = `Wrote ${filepath} (${content.length} chars)`;
    if (callback) await callback({ content: { text: msg } });
    return { success: true, text: msg };
  },
};

File: plugins/plugin-code/rust/src/actions/write_file.rs
use async_trait::async_trait;
use serde_json::Value;

use crate::service::CoderService;
use crate::{Action, ActionExample, ActionResult};

pub struct WriteFileAction;

#[async_trait]
impl Action for WriteFileAction {
    fn name(&self) -> &str {
        "WRITE_FILE"
    }

    fn similes(&self) -> Vec<&str> {
        vec!["CREATE_FILE", "SAVE_FILE", "OUTPUT_FILE"]
    }

    fn description(&self) -> &str {
        "Create or overwrite a file with given content."
    }

    async fn validate(&self, _message: &Value, _state: &Value) -> bool {
        true
    }

    async fn handler(
        &self,
        message: &Value,
        state: &Value,
        service: Option<&mut CoderService>,
    ) -> ActionResult {
        let Some(svc) = service else {
            return ActionResult {
                success: false,
                text: "Coder service is not available.".to_string(),
                data: None,
                error: Some("missing_service".to_string()),
            };
        };

        let filepath = state.get("filepath").and_then(|v| v.as_str()).unwrap_or("");
        let content = state.get("content").and_then(|v| v.as_str()).unwrap_or("");
        if filepath.trim().is_empty() {
            return ActionResult {
                success: false,
                text: "Missing filepath.".to_string(),
                data: None,
                error: Some("missing_filepath".to_string()),
            };
        }

        let conv = message
            .get("room_id")
            .and_then(|v| v.as_str())
            .or_else(|| message.get("agent_id").and_then(|v| v.as_str()))
            .unwrap_or("default");

        match svc.write_file(conv, filepath, content).await {
            Ok(()) => ActionResult {
                success: true,
                text: format!("Wrote {} ({} chars)", filepath, content.len()),
                data: None,
                error: None,
            },
            Err(err) => ActionResult {
                success: false,
                text: err,
                data: None,
                error: Some("write_failed".to_string()),
            },
        }
    }

    fn examples(&self) -> Vec<ActionExample> {
        vec![ActionExample {
            user_message: "write README.md".to_string(),
            agent_response: "Writing file…".to_string(),
        }]
    }
}
